crypto-serde/src/lib.rs

```diff
@@ -17,47 +17,182 @@ extern crate alloc;
 
 pub use serde;
 
-use serde::{ser, Serialize};
+use core::fmt;
+
+use serde::de::{Error, Expected, SeqAccess, Visitor};
+use serde::ser::SerializeTuple;
+use serde::{Deserializer, Serialize, Serializer};
 
 #[cfg(feature = "alloc")]
-use {
-    alloc::vec::Vec,
-    serde::de::{self, Deserialize, Error},
-};
+use {alloc::vec::Vec, serde::de::Deserialize};
 
 #[cfg(feature = "zeroize")]
 use zeroize::Zeroize;
 
 /// Serialize the given type as lower case hex when using human-readable
 /// formats or binary if the format is binary.
-pub fn serialize_hex_lower_or_bin<S, T>(value: &T, serializer: S) -> Result<S::Ok, S::Error>
+pub fn serialize_array_hex_lower_or_bin<S, T>(value: &T, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+    T: AsRef<[u8]>,
+{
+    #[cfg(feature = "alloc")]
+    if serializer.is_human_readable() {
+        return base16ct::lower::encode_string(value.as_ref()).serialize(serializer);
+    }
+    #[cfg(not(feature = "alloc"))]
+    if serializer.is_human_readable() {
+        return Err(S::Error::custom(
+            "serializer is human readable, which requires the `alloc` crate feature",
+        ));
+    }
+
+    let mut seq = serializer.serialize_tuple(value.as_ref().len())?;
+
+    for byte in value.as_ref() {
+        seq.serialize_element(byte)?;
+    }
+
+    seq.end()
+}
+
+/// Serialize the given type as upper case hex when using human-readable
+/// formats or binary if the format is binary.
+pub fn serialize_array_hex_upper_or_bin<S, T>(value: &T, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+    T: AsRef<[u8]>,
+{
+    #[cfg(feature = "alloc")]
+    if serializer.is_human_readable() {
+        return base16ct::upper::encode_string(value.as_ref()).serialize(serializer);
+    }
+    #[cfg(not(feature = "alloc"))]
+    if serializer.is_human_readable() {
+        return Err(S::Error::custom(
+            "serializer is human readable, which requires the `alloc` crate feature",
```

Comment:
There's not much point providing an error message in this case. serde will just throw it away:

https://github.com/serde-rs/serde/blob/38edb47/serde/src/de/value.rs#L73-L81

Also it'd probably be good to extract the `base16ct::upper::encode_string(value.as_ref()).serialize(serializer)` bits into a function (e.g. `serialize_hex`) which is conditionally defined when alloc is available and returns an error when it isn't, rather than duplicating this logic.

Reply:
That's just the default implementation, bincode actually does something:
https://docs.rs/bincode/1.3.3/src/bincode/error.rs.html#112-114

Reply:
Looks like `bincode` is a std-only crate?

I guess it works if the format implementation has alloc enabled but `crypto-serde` (which is `alloc` on-by-default) has it disabled.

Maybe extract the error message into a constant then.

Reply:
I already abstracted it into a single function, there should be no code duplication around this now. Tell me if what you think of it:
https://github.com/RustCrypto/formats/pull/538/commits/21de019aacd21af282e58b14e75284fcec26e40f

Reply:
Looks good 👍